File: apps/claude-plugin/cmd/hooks/load_skills.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
)

// LoadSkillsInput represents the input for load-skills command
type LoadSkillsInput struct {
	Skills    []string `json:"skills,omitempty"`
	Scenario  string   `json:"scenario,omitempty"`
	SkillsDir string   `json:"skillsDir,omitempty"`
}

// LoadSkillsOutput represents the output for load-skills command
type LoadSkillsOutput struct {
	Success     bool              `json:"success"`
	Content     string            `json:"content,omitempty"`
	Skills      []SkillInfo       `json:"skills,omitempty"`
	SkillsDir   string            `json:"skillsDir,omitempty"`
	Error       string            `json:"error,omitempty"`
	FilesLoaded []string          `json:"filesLoaded,omitempty"`
}

// SkillInfo represents information about a loaded skill
type SkillInfo struct {
	Name    string `json:"name"`
	Path    string `json:"path"`
	Content string `json:"content,omitempty"`
}

// RunLoadSkills handles the load-skills command
// It loads skill markdown files from the skills/ directory
func RunLoadSkills() error {
	// Read input from stdin (optional)
	input, _ := io.ReadAll(os.Stdin)

	var loadInput LoadSkillsInput
	if len(input) > 0 {
		// Try to parse as JSON
		json.Unmarshal(input, &loadInput)
	}

	// Resolve skills directory
	skillsDir := loadInput.SkillsDir
	if skillsDir == "" {
		skillsDir = resolveSkillsPath()
	}

	if skillsDir == "" {
		return outputJSON(LoadSkillsOutput{
			Success: false,
			Error:   "Could not locate skills directory",
		})
	}

	output := LoadSkillsOutput{
		Success:     true,
		SkillsDir:   skillsDir,
		FilesLoaded: []string{},
	}

	var contentBuilder strings.Builder

	// If specific skills requested, load only those
	if len(loadInput.Skills) > 0 {
		for _, skillName := range loadInput.Skills {
			skillContent, files, err := loadSkill(skillsDir, skillName)
			if err != nil {
				// Continue with other skills
				continue
			}
			output.Skills = append(output.Skills, SkillInfo{
				Name:    skillName,
				Path:    filepath.Join(skillsDir, skillName),
				Content: skillContent,
			})
			output.FilesLoaded = append(output.FilesLoaded, files...)
			contentBuilder.WriteString(skillContent)
			contentBuilder.WriteString("\n\n---\n\n")
		}
	} else if loadInput.Scenario != "" {
		// Load skills relevant to scenario
		skills := getSkillsForScenario(loadInput.Scenario)
		for _, skillName := range skills {
			skillContent, files, err := loadSkill(skillsDir, skillName)
			if err != nil {
				continue
			}
			output.Skills = append(output.Skills, SkillInfo{
				Name:    skillName,
				Path:    filepath.Join(skillsDir, skillName),
				Content: skillContent,
			})
			output.FilesLoaded = append(output.FilesLoaded, files...)
			contentBuilder.WriteString(skillContent)
			contentBuilder.WriteString("\n\n---\n\n")
		}
	} else {
		// Load all skills
		entries, err := os.ReadDir(skillsDir)
		if err != nil {
			return outputJSON(LoadSkillsOutput{
				Success: false,
				Error:   fmt.Sprintf("Failed to read skills directory: %v", err),
			})
		}

		for _, entry := range entries {
			if !entry.IsDir() {
				continue
			}
			skillName := entry.Name()
			skillContent, files, err := loadSkill(skillsDir, skillName)
			if err != nil {
				continue
			}
			output.Skills = append(output.Skills, SkillInfo{
				Name:    skillName,
				Path:    filepath.Join(skillsDir, skillName),
				Content: skillContent,
			})
			output.FilesLoaded = append(output.FilesLoaded, files...)
			contentBuilder.WriteString(skillContent)
			contentBuilder.WriteString("\n\n---\n\n")
		}
	}

	output.Content = contentBuilder.String()

	return outputJSON(output)
}

// loadSkill loads a single skill and all its markdown files
func loadSkill(skillsDir, skillName string) (string, []string, error) {
	skillPath := filepath.Join(skillsDir, skillName)

	// Check if skill directory exists
	if _, err := os.Stat(skillPath); os.IsNotExist(err) {
		return "", nil, fmt.Errorf("skill not found: %s", skillName)
	}

	var contentBuilder strings.Builder
	var filesLoaded []string

	// Load main SKILL.md first
	mainSkillFile := filepath.Join(skillPath, "SKILL.md")
	if content, err := os.ReadFile(mainSkillFile); err == nil {
		contentBuilder.WriteString(fmt.Sprintf("# Skill: %s\n\n", skillName))
		contentBuilder.Write(content)
		contentBuilder.WriteString("\n\n")
		filesLoaded = append(filesLoaded, mainSkillFile)
	}

	// Load other top-level markdown files
	topLevelFiles := []string{"ANALYSIS.md", "PLANNING.md", "CODING.md"}
	for _, fileName := range topLevelFiles {
		filePath := filepath.Join(skillPath, fileName)
		if content, err := os.ReadFile(filePath); err == nil {
			contentBuilder.WriteString(fmt.Sprintf("## %s\n\n", strings.TrimSuffix(fileName, ".md")))
			contentBuilder.Write(content)
			contentBuilder.WriteString("\n\n")
			filesLoaded = append(filesLoaded, filePath)
		}
	}

	// Load scenarios subdirectory
	scenariosDir := filepath.Join(skillPath, "scenarios")
	if entries, err := os.ReadDir(scenariosDir); err == nil {
		for _, entry := range entries {
			if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".md") {
				continue
			}
			filePath := filepath.Join(scenariosDir, entry.Name())
			if content, err := os.ReadFile(filePath); err == nil {
				scenarioName := strings.TrimSuffix(entry.Name(), ".md")
				contentBuilder.WriteString(fmt.Sprintf("## Scenario: %s\n\n", scenarioName))
				contentBuilder.Write(content)
				contentBuilder.WriteString("\n\n")
				filesLoaded = append(filesLoaded, filePath)
			}
		}
	}

	// Load templates subdirectory
	templatesDir := filepath.Join(skillPath, "templates")
	if entries, err := os.ReadDir(templatesDir); err == nil {
		contentBuilder.WriteString("## Templates\n\n")
		for _, entry := range entries {
			if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".md") {
				continue
			}
			filePath := filepath.Join(templatesDir, entry.Name())
			if content, err := os.ReadFile(filePath); err == nil {
				templateName := strings.TrimSuffix(entry.Name(), ".md")
				contentBuilder.WriteString(fmt.Sprintf("### Template: %s\n\n", templateName))
				contentBuilder.Write(content)
				contentBuilder.WriteString("\n\n")
				filesLoaded = append(filesLoaded, filePath)
			}
		}
	}

	return contentBuilder.String(), filesLoaded, nil
}

// getSkillsForScenario returns the skills relevant to a scenario
func getSkillsForScenario(scenario string) []string {
	scenarioSkills := map[string][]string{
		"ANALYSIS": {"coding-workflow", "writing-notes"},
		"RESEARCH": {"writing-notes"},
		"DECISION": {"writing-notes"},
		"FEATURE":  {"coding-workflow", "writing-notes"},
		"SPEC":     {"coding-workflow", "writing-notes"},
		"TESTING":  {"coding-workflow", "writing-notes"},
		"BUG":      {"coding-workflow", "writing-notes"},
	}

	if skills, ok := scenarioSkills[strings.ToUpper(scenario)]; ok {
		return skills
	}

	// Default to all skills
	return []string{"coding-workflow", "writing-notes"}
}

File: hooks/scripts/load-skills.ts
/**
 * Load-skills command handler.
 *
 * Ported from apps/claude-plugin/cmd/hooks/load_skills.go.
 * Loads skill markdown files from the skills/ directory.
 */
import { readFileSync, readdirSync, statSync } from "node:fs";
import { join, dirname } from "node:path";
import type { LoadSkillsInput, LoadSkillsOutput, SkillInfo } from "./types.js";

/** Map of scenarios to relevant skill names. */
const scenarioSkills: Record<string, string[]> = {
  ANALYSIS: ["coding-workflow", "writing-notes"],
  RESEARCH: ["writing-notes"],
  DECISION: ["writing-notes"],
  FEATURE: ["coding-workflow", "writing-notes"],
  SPEC: ["coding-workflow", "writing-notes"],
  TESTING: ["coding-workflow", "writing-notes"],
  BUG: ["coding-workflow", "writing-notes"],
};

/** Get the plugin root directory. */
function getPluginRoot(): string {
  const envRoot = process.env.CLAUDE_PLUGIN_ROOT;
  if (envRoot) return envRoot;

  // Fall back to two levels up from this script
  // Script is at hooks/scripts/load-skills.ts, root is ../../
  return dirname(dirname(dirname(new URL(import.meta.url).pathname)));
}

/** Resolve the skills directory path. */
function resolveSkillsPath(): string {
  const root = getPluginRoot();
  if (!root) {
    try {
      statSync("skills");
      return "skills";
    } catch {
      return "";
    }
  }
  return join(root, "skills");
}

/** Load a single skill and all its markdown files. */
function loadSkill(
  skillsDir: string,
  skillName: string,
): { content: string; files: string[] } | null {
  const skillPath = join(skillsDir, skillName);
  try {
    statSync(skillPath);
  } catch {
    return null;
  }

  const parts: string[] = [];
  const filesLoaded: string[] = [];

  // Load main SKILL.md first
  const mainFile = join(skillPath, "SKILL.md");
  try {
    const content = readFileSync(mainFile, "utf-8");
    parts.push(`# Skill: ${skillName}\n\n${content}\n\n`);
    filesLoaded.push(mainFile);
  } catch {
    // No SKILL.md
  }

  // Load other top-level markdown files
  for (const fileName of ["ANALYSIS.md", "PLANNING.md", "CODING.md"]) {
    const filePath = join(skillPath, fileName);
    try {
      const content = readFileSync(filePath, "utf-8");
      parts.push(`## ${fileName.replace(".md", "")}\n\n${content}\n\n`);
      filesLoaded.push(filePath);
    } catch {
      // File doesn't exist
    }
  }

  // Load scenarios subdirectory
  const scenariosDir = join(skillPath, "scenarios");
  try {
    const entries = readdirSync(scenariosDir);
    for (const entry of entries) {
      if (!entry.endsWith(".md")) continue;
      const filePath = join(scenariosDir, entry);
      try {
        if (!statSync(filePath).isFile()) continue;
        const content = readFileSync(filePath, "utf-8");
        const name = entry.replace(".md", "");
        parts.push(`## Scenario: ${name}\n\n${content}\n\n`);
        filesLoaded.push(filePath);
      } catch {
        // Skip
      }
    }
  } catch {
    // No scenarios dir
  }

  // Load templates subdirectory
  const templatesDir = join(skillPath, "templates");
  try {
    const entries = readdirSync(templatesDir);
    if (entries.some((e) => e.endsWith(".md"))) {
      parts.push("## Templates\n\n");
      for (const entry of entries) {
        if (!entry.endsWith(".md")) continue;
        const filePath = join(templatesDir, entry);
        try {
          if (!statSync(filePath).isFile()) continue;
          const content = readFileSync(filePath, "utf-8");
          const name = entry.replace(".md", "");
          parts.push(`### Template: ${name}\n\n${content}\n\n`);
          filesLoaded.push(filePath);
        } catch {
          // Skip
        }
      }
    }
  } catch {
    // No templates dir
  }

  return { content: parts.join(""), files: filesLoaded };
}

/** Get skills relevant to a scenario. */
function getSkillsForScenario(scenario: string): string[] {
  return scenarioSkills[scenario.toUpperCase()] ?? [
    "coding-workflow",
    "writing-notes",
  ];
}

export async function runLoadSkills(): Promise<void> {
  const chunks: Buffer[] = [];
  for await (const chunk of process.stdin) {
    chunks.push(chunk as Buffer);
  }
  const raw = Buffer.concat(chunks).toString("utf-8").trim();

  let loadInput: LoadSkillsInput = {};
  if (raw) {
    try {
      loadInput = JSON.parse(raw) as LoadSkillsInput;
    } catch {
      // Ignore
    }
  }

  // Resolve skills directory
  let skillsDir = loadInput.skillsDir || resolveSkillsPath();
  if (!skillsDir) {
    process.stdout.write(
      JSON.stringify(
        { success: false, error: "Could not locate skills directory" },
        null,
        2,
      ) + "\n",
    );
    return;
  }

  const output: LoadSkillsOutput = {
    success: true,
    skillsDir,
    filesLoaded: [],
  };

  const contentParts: string[] = [];
  const skills: SkillInfo[] = [];

  // Determine which skills to load
  let skillNames: string[];

  if (loadInput.skills && loadInput.skills.length > 0) {
    skillNames = loadInput.skills;
  } else if (loadInput.scenario) {
    skillNames = getSkillsForScenario(loadInput.scenario);
  } else {
    // Load all skills
    try {
      skillNames = readdirSync(skillsDir).filter((entry) => {
        try {
          return statSync(join(skillsDir, entry)).isDirectory();
        } catch {
          return false;
        }
      });
    } catch {
      process.stdout.write(
        JSON.stringify(
          {
            success: false,
            error: `Failed to read skills directory: ${skillsDir}`,
          },
          null,
          2,
        ) + "\n",
      );
      return;
    }
  }

  for (const skillName of skillNames) {
    const result = loadSkill(skillsDir, skillName);
    if (!result) continue;

    skills.push({
      name: skillName,
      path: join(skillsDir, skillName),
      content: result.content,
    });
    output.filesLoaded!.push(...result.files);
    contentParts.push(result.content);
    contentParts.push("\n\n---\n\n");
  }

  output.skills = skills;
  output.content = contentParts.join("");

  process.stdout.write(JSON.stringify(output, null, 2) + "\n");
}

// Run if executed directly
const isMain = process.argv[1]?.endsWith("load-skills.ts") ||
  process.argv[1]?.endsWith("load-skills.js");
if (isMain) {
  runLoadSkills().catch((err) => {
    process.stderr.write(`Error: ${err}\n`);
    process.exit(1);
  });
}
